File: bindings/lua/proven/doh.lua
-- SPDX-License-Identifier: PMPL-1.0-or-later
-- Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
--
-- DoH protocol types for proven-servers.

local M = {}

--- ContentType matching the Idris2 ABI tags.
M.ContentType = {
    DNS_MESSAGE = 0,
    DNS_JSON = 1,
}

--- RequestMethod matching the Idris2 ABI tags.
M.RequestMethod = {
    GET = 0,
    POST = 1,
}

--- WireFormat matching the Idris2 ABI tags.
M.WireFormat = {
    BINARY = 0,
    JSON = 1,
}

--- ErrorReason matching the Idris2 ABI tags.
M.ErrorReason = {
    BAD_CONTENT_TYPE = 0,
    BAD_METHOD = 1,
    PAYLOAD_TOO_LARGE = 2,
    UPSTREAM_TIMEOUT = 3,
    UPSTREAM_ERROR = 4,
}

--- SessionState matching the Idris2 ABI tags.
M.SessionState = {
    IDLE = 0,
    BOUND = 1,
    SERVING = 2,
    RESOLVING = 3,
    SHUTDOWN = 4,
}

return M

File: bindings/go/doh.go
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>

// DoH protocol types for proven-servers.
// All tag values match the Idris2 ABI discriminants exactly.
package proven

// ContentType represents the ContentType type (Idris2 ABI tags).
type ContentType uint8

const (
	ContentTypeDnsMessage ContentType = iota
	ContentTypeDnsJson
)

// RequestMethod represents the RequestMethod type (Idris2 ABI tags).
type RequestMethod uint8

const (
	RequestMethodGet RequestMethod = iota
	RequestMethodPost
)

// WireFormat represents the WireFormat type (Idris2 ABI tags).
type WireFormat uint8

const (
	WireFormatBinary WireFormat = iota
	WireFormatJson
)

// ErrorReason represents the ErrorReason type (Idris2 ABI tags).
type ErrorReason uint8

const (
	ErrorReasonBadContentType ErrorReason = iota
	ErrorReasonBadMethod
	ErrorReasonPayloadTooLarge
	ErrorReasonUpstreamTimeout
	ErrorReasonUpstreamError
)

// SessionState represents the SessionState type (Idris2 ABI tags).
type SessionState uint8

const (
	SessionStateIdle SessionState = iota
	SessionStateBound
	SessionStateServing
	SessionStateResolving
	SessionStateShutdown
)
